### PR TITLE
Add support for Dj40, drop Py36 and Dj31

### LINKED ISSUE #1037: Django 4.0 compatibility
**Describe the bug**
When used with Django 4.0, there are reports of a missing migration. For example:
```
Migrations for 'oauth2_provider':
  /usr/local/lib/python3.8/dist-packages/oauth2_provider/migrations/0005_alter_accesstoken_user_alter_application_user_and_more.py
    - Alter field user on accesstoken
    - Alter field user on application
    - Alter field user on grant
    - Alter field user on idtoken
    - Alter field user on refreshtoken
```
By default, of course, normal users cannot write such a file, and applying migrations without causes warnings like this:
```
 Your models in app(s): 'oauth2_provider' have changes that are not yet reflected in a migration, and so won't be applied.
  Run 'manage.py makemigrations' to make new migrations, and then re-run 'manage.py migrate' to apply them.
```
I note that Django 4.0 release notes mention the possibility of some no-op migrations being needed...I am not sure if this is the root cause.

**To Reproduce**
Run "./manage.py makemigrations" in a project which uses this package.

**Expected behavior**
Any needed migration should be provided.

**Version**
- [x] I have tested with the latest published release and it's still a problem.
- [ ] I have tested with the master branch and it's still a problem.

**Additional context**

### PR DESCRIPTION
**Pull request by Andrew-Chen-Wang**
Fixes #1037 

## Description of the Change

Add support for Dj40, drop Py36 and Dj31

Dj31 and Py36 will be EOL end of this December of 2021. I do not want to hold back any PRs that need to support old versions of stuff, thus I have dropped those two.

## Checklist

- [X] PR only contains one change (considered splitting up PR)
- [ ] unit-test added
- [ ] documentation updated
- [ ] `CHANGELOG.md` updated (only for user relevant changes)
- [X] author name in `AUTHORS`

### CODE REVIEW

**Comment by Andrew-Chen-Wang** [comment]
Note: do not merge this as of yet. I'll need to run the quick migration.

**Comment by Andrew-Chen-Wang** [comment]
@n2ygk In the settings, you'll need to remove the required check for Python 3.6 due to the dropping of Python 3.6 in this PR. Thanks!

**Comment by Andrew-Chen-Wang** [comment]
I believe there is a bug in Django which is mentioned in #1037. I'm not sure how to resolve it, but it seems like an unintended bug in Django itself, not our end. I've filed a bug (which will probably get closed wontfix) https://code.djangoproject.com/ticket/33366

@auvipy I will be taking finals for next 3 days, so please take over. Cheers :)

File: .github/workflows/test.yml
Comment (auvipy) [inline .github/workflows/test.yml:12]: good change

**Review by auvipy — COMMENTED** [review]
ok will check what remains

**Comment by auvipy** [comment]
I think python 3.10 is not backported in django 3.2?

**Comment by auvipy** [comment]
build (3.6) Expected — Waiting for status to be reported what is this build?

**Comment by Andrew-Chen-Wang** [comment]
I removed Py36. Authors in settings can require certain builds to be successful before PR merge. Authors should disable this. If they don't then we can add Python 3.6 back in to circumvent.

**Comment by auvipy** [comment]
I am not really sure why python 3.10 is failing

**Review by auvipy — APPROVED** [review]

**Comment by Andrew-Chen-Wang** [comment]
Please do not merge until the migration file is properly fixed.

**Comment by auvipy** [comment]
we actually cant merge this due to required check. also the migration issue is not actual issue. thats due to django new settings/features as far as I know.

**Comment by Andrew-Chen-Wang** [comment]
Yes, however, like Django 3.2, users kept creating migration files inside the package itself since people we're using virtual envs. This is unintended behavior and can cause huge issues down the road when adding new migration files as migration file IDs will be duplicated due to an accidental  dev-side migration. 

TL;DR if we don't fix the migration file issue, people's production servers will crash.

**Comment by Andrew-Chen-Wang** [comment]
@auvipy Python 3.10 failing due to main branch of Django. Somehow `sqlite3.OperationalError: Cannot add a UNIQUE column` this error is now popping up. Must mean either a bug or something new that we're not aware of.

@n2ygk @jezdez  In the settings, you'll need to remove the required check for Python 3.6 due to the dropping of Python 3.6 in this PR. Thanks!

**Comment by auvipy** [comment]
as I said this seems to be a regression in django https://github.com/django/django/pull/15205/files

**Comment by auvipy** [comment]
Also I don't think it's a blocker for release. we can drop this after a new release

**Comment by Andrew-Chen-Wang** [comment]
Sounds good!

**Comment by auvipy** [comment]
not a problem, we can do 1.6.1 soon :D

**Comment by jezdez** [comment]
> @n2ygk @jezdez In the settings, you'll need to remove the required check for Python 3.6 due to the dropping of Python 3.6 in this PR. Thanks!

Done that.

**Comment by jezdez** [comment]
> Also I don't think it's a blocker for release. we can drop this after a new release

FWIW dropping support for a Django or Python version shouldn't be done in a patch version, I'd suggest to do a 1.7.0 with this change applied.

File: tox.ini
Comment (Andrew-Chen-Wang) [inline tox.ini:6]: Should include 3.10 since dj 3.2.9 added support

**Review by Andrew-Chen-Wang — COMMENTED** [review]
@jezdez I think this should be merged though with a change in setup.py to disallow Django 4.0.0 but allow 4.0.1. That way it'd be included in this new minor release as it seems like current maintainers are fairly busy.

We can just delete the current git tag and re tag since not released on pypi yet

**Comment by Andrew-Chen-Wang** [comment]
ah nvm, didn't see that it was released 😅

File: tox.ini
Comment (jezdez) [reply tox.ini:6]: > @jezdez I think this should be merged though with a change in setup.py to disallow Django 4.0.0 but allow 4.0.1. That way it'd be included in this new minor release as it seems like current maintainers are fairly busy.

That's...not really a great way of specifying dependencies. 

@auvipy Please make sure to coordinate with @n2ygk and @MattBlack85 before tagging. We've talked about this before, you can't just ask for review in https://github.com/jazzband/django-oauth-toolkit/issues/1000#issuecomment-995895750 and then go ahead and create a tag on your own. Drafting a release is one thing, but publishing the release here is another.

> 
> We can just delete the current git tag and re tag since not released on pypi yet

Sure, just know that every git tag triggers the release process and sends an email to the project leads to review it, so coordinating this beforehand would be better.

Comment (auvipy) [reply tox.ini:6]: > > @jezdez I think this should be merged though with a change in setup.py to disallow Django 4.0.0 but allow 4.0.1. That way it'd be included in this new minor release as it seems like current maintainers are fairly busy.
> 
> That's...not really a great way of specifying dependencies.
> 
> @auvipy Please make sure to coordinate with @n2ygk and @MattBlack85 before tagging. We've talked about this before, you can't just ask for review in [#1000 (comment)](https://github.com/jazzband/django-oauth-toolkit/issues/1000#issuecomment-995895750) and then go ahead and create a tag on your own. Drafting a release is one thing, but publishing the release here is another.
> 
> > We can just delete the current git tag and re tag since not released on pypi yet
> 
> Sure, just know that every git tag triggers the release process and sends an email to the project leads to review it, so coordinating this beforehand would be better.

I haven't seen any guideline for that, and people was bugging me for drafting a release. they were not available for long time if you check. as the github tag push is for review I thought it is drafting the release. hence the confusion. I think toy should specify loudly to people that not all regular contributors are project lead or able to push e tag. this type of mixed situation in a same or is kind of frustrating & de-motivating. I think it's wise to not involve in any jazzband projects without being a project lead.

Comment (Andrew-Chen-Wang) [reply tox.ini:6]: @jezdez can you recommend a way to pin dependencies? I think allowing django-oauth-toolkit run on Django 4.0.0 will cause the same mass production failures that Django 3.2 brought with default auto field. I was just thinking `django<4.0.0,>=4.0.1` (sorry if that's not the right syntax).

Or are you saying don't pin at all? Can you elaborate since I'm not quite understanding 😅

Comment (jezdez) [reply tox.ini:6]: > > > @jezdez I think this should be merged though with a change in setup.py to disallow Django 4.0.0 but allow 4.0.1. That way it'd be included in this new minor release as it seems like current maintainers are fairly busy.
> > 
> > 
> > That's...not really a great way of specifying dependencies.
> > @auvipy Please make sure to coordinate with @n2ygk and @MattBlack85 before tagging. We've talked about this before, you can't just ask for review in [#1000 (comment)](https://github.com/jazzband/django-oauth-toolkit/issues/1000#issuecomment-995895750) and then go ahead and create a tag on your own. Drafting a release is one thing, but publishing the release here is another.
> > > We can just delete the current git tag and re tag since not released on pypi yet
> > 
> > 
> > Sure, just know that every git tag triggers the release process and sends an email to the project leads to review it, so coordinating this beforehand would be better.
> 
> I haven't seen any guideline for that, and people was bugging me for drafting a release. they were not available for long time if you check. as the github tag push is for review I thought it is drafting the release. hence the confusion.

I hear you, can you elaborate what you mean that the project leads haven't been available for a long time? Who was bugging you to do a release?

To clarify what "drafting a release" means: that's for example updating the changelog, reviewing any potentially needed PRs like this one and finally either open a PR asking the project leads to review it or (as you chose) to create the release. I've gotten word from the project lead @n2ygk that he didn't not expect the release and would have happily done code reviews if being requested to do one.

> I think toy should specify loudly to people that not all regular contributors are project lead or able to push e tag. this type of mixed situation in a same or is kind of frustrating & de-motivating. I think it's wise to not involve in any jazzband projects without being a project lead.

But @MattBlack85 and @n2ygk are the project leads for this project, I'm not sure I follow your argument.

Comment (jezdez) [reply tox.ini:6]: > @jezdez can you recommend a way to pin dependencies? I think allowing django-oauth-toolkit run on Django 4.0.0 will cause the same mass production failures that Django 3.2 brought with default auto field. I was just thinking `django<4.0.0,>=4.0.1` (sorry if that's not the right syntax).
> 
> Or are you saying don't pin at all? Can you elaborate since I'm not quite understanding 😅

I'm not sure I follow, is this about https://github.com/django/django/pull/15205? If Django 4.0.0 isn't possible to run properly, then it should be excluded with the [dependency spec](https://www.python.org/dev/peps/pep-0508/), with the `!=4.0.0` notation. That will prevent the installation of 4.0.0 at least and would create a installation error for a user who would pin to Django 4.0.0 in one of their other project dependencies.

Comment (Andrew-Chen-Wang) [reply tox.ini:6]: @jezdez it was in response to the quote below, and yes your suggestion was what I was thinking as a way to avoid installation of Django 4.0.0 and django-oauth-toolkit.

>> @jezdez I think this should be merged though with a change in setup.py to disallow Django 4.0.0 but allow 4.0.1. That way it'd be included in this new minor release as it seems like current maintainers are fairly busy.
> That's...not really a great way of specifying dependencies.

Comment (n2ygk) [reply tox.ini:6]: @Andrew-Chen-Wang I've pushed a commit that makes the minimum dj40 be 4.0.1 but I see from your comments that there are other concerns that may make this not yet ready to merge? I'll mark this PR WIP to make sure it doesn't get merged just yet.

Comment (Andrew-Chen-Wang) [reply tox.ini]: @n2ygk the docs should now be updated. I've also added the !=4.0.0 constraint to setup.cfg to prevent installation. Let me know if that's not a good practice.

**Comment by n2ygk** [comment]
@Andrew-Chen-Wang oops you and I crossed commits in the mail... I think the tox should be what I had if my understanding is correct:
```diff
-    dj40: Django>=4.0.1,<4.1
+    dj40: Django>=4.0.0,<4.1
```

**Comment by Andrew-Chen-Wang** [comment]
Nope! Django 4.0.1 isn't released yet, but I do want to test 4.0. Please read my explanation above 😅 (sorry edited it awhile ago).

**Comment by n2ygk** [comment]
> Nope! Django 4.0.1 isn't released yet, but I do want to test 4.0. Please read my explanation above 😅 (sorry edited it awhile ago).

D'oh!

Looks like everything but py310 passed the tests.

**Comment by Andrew-Chen-Wang** [comment]
Hmm that's why auvipy removed it due to Django main failing on Python 3.10. I personally don't understand what's going on 😅, but what I do know now is that it's also failing on Python 3.9 and others, but Too successfully ignored those ones but didn't properly ignore Python 3.10. I'll get tox to ignore the errors then.

Also, do you mind requiring Python 3.10 in the settings? Thanks!

**Comment by Andrew-Chen-Wang** [comment]
This should be the last PR, so it'll require deleting the 1.6.0 git tag release and recreating it to include this PR. Also enabling Python 3.10 requirement in CI in the GitHub settings. Finally, you'll be able to make the Jazzband release. Thanks for a great library!

**Comment by n2ygk** [comment]
Looks like a failure in 3.9 now so will need to investigate that.

**Comment by n2ygk** [comment]
> Looks like a failure in 3.9 now so will need to investigate that.

Hmm. ran tox locally and 3.9 was OK:
```
ERROR:   flake8: commands failed
  docs: commands succeeded
  py37-dj22: commands succeeded
  py38-dj22: commands succeeded
  py39-dj22: commands succeeded
  py37-dj32: commands succeeded
  py38-dj32: commands succeeded
  py39-dj32: commands succeeded
  py310-dj32: commands succeeded
  py38-dj40: commands succeeded
  py39-dj40: commands succeeded
  py310-dj40: commands succeeded
  py38-djmain: ignored failed command
  py39-djmain: ignored failed command
  py310-djmain: ignored failed command
```

**Comment by n2ygk** [comment]
I took the liberty of using this PR to update the CHANGELOG to be hopefully more user-friendly. We don't need to list every single PR and issue, only those that have an impact on a user considering upgrading.

**Review by n2ygk — APPROVED** [review]

**Comment by n2ygk** [comment]
Reopening this until we get to an acceptable solution with Django >= 4.0 so we don't lose track. Target the next milestone....

**Pull request merged**
